File: mobile/examples/swift/hello_world/DemoFilter.swift
import Envoy
import Foundation

/// Example of a simple HTTP filter that adds a response header.
struct DemoFilter: ResponseFilter {
  func onResponseHeaders(_ headers: ResponseHeaders, endStream: Bool, streamIntel: StreamIntel)
    -> FilterHeadersStatus<ResponseHeaders>
  {
    let builder = headers.toResponseHeadersBuilder()
    builder.add(name: "filter-demo", value: "1")
    return .continue(headers: builder.build())
  }

  func setResponseFilterCallbacks(_ callbacks: ResponseFilterCallbacks, streamIntel: StreamIntel) {}

  func onResponseData(_ body: Data, endStream: Bool, streamIntel: StreamIntel)
    -> FilterDataStatus<ResponseHeaders>
  {
    return .continue(data: body)
  }

  func onResponseTrailers(_ trailers: ResponseTrailers, streamIntel: StreamIntel)
    -> FilterTrailersStatus<ResponseHeaders, ResponseTrailers>
  {
    return .continue(trailers: trailers)
  }

  func onError(_ error: EnvoyError, streamIntel: StreamIntel) {}

  func onCancel(streamIntel: StreamIntel) {}
}

File: mobile/examples/kotlin/hello_world/DemoFilter.kt
package io.envoyproxy.envoymobile.helloenvoykotlin

import android.util.Log
import io.envoyproxy.envoymobile.EnvoyError
import io.envoyproxy.envoymobile.FilterDataStatus
import io.envoyproxy.envoymobile.FilterHeadersStatus
import io.envoyproxy.envoymobile.FilterTrailersStatus
import io.envoyproxy.envoymobile.ResponseFilter
import io.envoyproxy.envoymobile.ResponseHeaders
import io.envoyproxy.envoymobile.ResponseTrailers
import io.envoyproxy.envoymobile.StreamIntel
import java.nio.ByteBuffer

class DemoFilter : ResponseFilter {
  override fun onResponseHeaders(
    headers: ResponseHeaders,
    endStream: Boolean,
    streamIntel: StreamIntel
  ): FilterHeadersStatus<ResponseHeaders> {
    Log.d("DemoFilter", "On headers!")
    val builder = headers.toResponseHeadersBuilder()
    builder.add("filter-demo", "1")
    return FilterHeadersStatus.Continue(builder.build())
  }

  override fun onResponseData(
    body: ByteBuffer,
    endStream: Boolean,
    streamIntel: StreamIntel
  ): FilterDataStatus<ResponseHeaders> {
    Log.d("DemoFilter", "On data!")
    return FilterDataStatus.Continue(body)
  }

  override fun onResponseTrailers(
    trailers: ResponseTrailers,
    streamIntel: StreamIntel
  ): FilterTrailersStatus<ResponseHeaders, ResponseTrailers> {
    Log.d("DemoFilter", "On trailers!")
    return FilterTrailersStatus.Continue(trailers)
  }

  override fun onError(error: EnvoyError, streamIntel: StreamIntel) {
    Log.d("DemoFilter", "On error!")
  }

  override fun onCancel(streamIntel: StreamIntel) {
    Log.d("DemoFilter", "On cancel!")
  }
}
